### PR TITLE
Remove QObjects from ALCFitting Presenter and Model

### LINKED ISSUE #38007: ALCPeakFittingPresenter inherits from a QObject
**Describe the outcome that is desired.**
The Muon ALC interface is comprised of three pages. Each page as a separate MVP pattern. One of the pages is known as peak fitting, and uses the classes `ALCPeakFittingView`, `ALCPeakFittingModel` and `ALCPeakFittingPresenter`.

The ALCPeakFittingPresenter class inherits from a QObject (meaning Qt is being used in the presenter). We do not want the Presenter to use Qt because it makes it difficult to test the presenter and violates the separation of concerns. All Qt code should be isolated to the View.

The Presenter being a QObject suggests that there are signals from the view being connected to slots in the presenter. To avoid making these cross-boundary connections, you can subscribe a non-owning raw pointer of the presenter to the view. The view can then call a method in the presenter whenever an even occurs.

To make it possible to mock the View in our tests for the presenter, we probably want to make use of an interface class. This allows us to swap out the view easily in our tests with a "MockView". It appears the interface class for the view DOES exist, however it inherits from a QObject - which negates the whole point of having an interface class! The interface class for the view should not inherit from a QObject.

**Describe any solutions you are considering**
We want to remove all Qt code from the Presenter. To do this we should:
- Subscribe the presenter to the view, and use callbacks to the presenter whenever an event occurs
- Use an interface class for the view, so that we can easily mock the view in the future

**Testing instructions**
You can use the testing instructions in this issue for an overview on how to use the Muon ALC interface:
https://github.com/mantidproject/mantid/pull/25879

**Acceptance criteria**
- [ ] The Presenter contains no Qt code
- [ ] The View inherits from an "Interface class" which is NOT a QObject

### PR DESCRIPTION
**Pull request by cailafinn**
#### Summary of work

Stops the presenter and the model from being QObjects, replacing their signals and slots with functionality based on the observer pattern where the presenter (as an observer) subscribes to the view and the model and is updated on their progress. 

As part of this, the view and model no longer directly communicate, making this compliant with MVP principles. 

Fixes #38007 

#### Further detail of work
Tests have been updated with new mocks to test the implemented functionality independently of any changes to the other parts of the MVP structure. 

### To test:
1. Run through the tests here: https://developer.mantidproject.org/Testing/MuonAnalysis_test_guides/Muon_ALC.html#alc-peak-fitting

*This does not require release notes* because **no user-facing functionality should have changed.**

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by adriazalvarez — CHANGES_REQUESTED** [review]
Trying to do the manual testing, there are two problems: 
On the baseline fitting step, there are three legends on the plot after fitting: 
Screenshot 2024-10-30 150329
And the data for the peak is not carried over to the peak fitting tab (it is empty): 
image

**Comment by cailafinn** [comment]
Thanks for the review @adriazalvarez 
Would you be able to provide the data you used for the test? I struggled to find it in any of the Usage/Sample data sets.

**Comment by cailafinn** [comment]
> And the data for the peak is not carried over to the peak fitting tab (it is empty)

This actually seems to be an issue on `main`, but I'll fix it as part of this PR.

File: qt/scientific_interfaces/Muon/ALCPeakFittingPresenter.h [hunk 1]
```diff
@@ -11,39 +11,34 @@
...
-namespace CustomInterfaces {
+namespace MantidQt::CustomInterfaces {
 
 /** ALCPeakFittingPresenter : Presenter for Peak Fitting step of ALC interface.
  */
-class MANTIDQT_MUONINTERFACE_DLL ALCPeakFittingPresenter : public QObject {
-  Q_OBJECT
+class MANTIDQT_MUONINTERFACE_DLL ALCPeakFittingPresenter : public IALCPeakFittingModelSubscriber,
+                                                           public IALCPeakFittingViewSubscriber {
+  //  Q_OBJECT
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCPeakFittingPresenter.h:24]: Can you remove the comment here?

File: qt/scientific_interfaces/Muon/ALCBaselineModellingPresenter.h [hunk 2]
```diff
@@ -42,6 +43,9 @@ class MANTIDQT_MUONINTERFACE_DLL ALCBaselineModellingPresenter : IALCBaselineMod
   /// Called when on of section selectors is modified
   void onSectionSelectorModified(int index) override;
 
+  /// Subscribe to updates on changes to the corrected data.
+  void subscribe(IALCBaselineModellingPresenterSubscriber *subscriber);
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCBaselineModellingPresenter.h:47]: Just asking your opinion, would it be better to rename this function as `subscribeToInterface` or `subscribeInterface`?

File: qt/scientific_interfaces/Muon/ALCInterface.cpp [hunk 3]
```diff
@@ -110,10 +110,14 @@ void ALCInterface::initLayout() {
   m_baselineModelling->initialize();
 
   m_peakFittingView = new ALCPeakFittingView(m_ui.peakFittingView);
-  m_peakFitting = new ALCPeakFittingPresenter(m_peakFittingView, m_peakFittingModel);
+  auto jobRunner = std::make_unique<MantidQt::API::QtJobRunner>(true);
+  auto algorithmRunner = std::make_unique<MantidQt::API::AlgorithmRunner>(std::move(jobRunner));
+  m_peakFittingModel = std::make_shared<ALCPeakFittingModel>(std::move(algorithmRunner));
+  m_peakFitting = new ALCPeakFittingPresenter(m_peakFittingView, m_peakFittingModel.get());
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCInterface.cpp:116]: could we store this peak fitting presenter as unique_ptr?

File: qt/scientific_interfaces/Muon/ALCPeakFittingModel.cpp [hunk 4]
```diff
@@ -7,15 +7,14 @@
...
 #include "MantidAPI/AlgorithmManager.h"
+#include "MantidAPI/AlgorithmRuntimeProps.h"
 #include "MantidAPI/CompositeFunction.h"
 #include "MantidAPI/ITableWorkspace.h"
 #include "MantidAPI/MatrixWorkspace.h"
 #include "MantidAPI/TableRow.h"
 #include "MantidAPI/TextAxis.h"
 #include "MantidAPI/WorkspaceFactory.h"
-
 #include <Poco/ActiveResult.h>
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCPeakFittingModel.cpp:17]: I think this include is not necessary anymore after adding the algorithm subscriber?

File: qt/scientific_interfaces/Muon/ALCPeakFittingView.cpp [hunk 5]
```diff
@@ -31,7 +33,7 @@ IPeakFunction_const_sptr ALCPeakFittingView::peakPicker() const { return m_peakP
 void ALCPeakFittingView::initialize() {
   m_ui.setupUi(m_widget);
 
-  connect(m_ui.fit, SIGNAL(clicked()), this, SIGNAL(fitRequested()));
+  connect(m_ui.fit, SIGNAL(clicked()), this, SLOT(fitRequested()));
```
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCPeakFittingView.cpp:36]: Can you implement the connections using the new qt syntax in signals/slots of this class?
```suggestion
  connect(m_ui.fit, &QPushButton::clicked, this, &ALCPeakFittingView::fitRequested);
```

**Review by adriazalvarez — CHANGES_REQUESTED** [review]
Tested the interface and everything works correctly on the peak fitting tab. 
It's a nice implementation of the MVP pattern with subscribers and is a clean and clear refactorization. I have found very small issues that in some cases are just questions. But it looks good otherwise. 
Also, I have checked(I didn't when I tested initially) that the recent refactorization for the MVP training of the baseline presenter was already on the code, and the issue I experienced initially was a bug with that, but thanks a lot for fixing it anyway!

File: qt/scientific_interfaces/Muon/ALCBaselineModellingPresenter.h (hunk 2 above)
Comment (cailafinn) [reply qt/scientific_interfaces/Muon/ALCBaselineModellingPresenter.h:47]: I think given the `subscribe` function is always called on the presenter, it provides enough context so can stay as is.

File: qt/scientific_interfaces/Muon/ALCPeakFittingView.cpp
Comment (adriazalvarez) [inline qt/scientific_interfaces/Muon/ALCPeakFittingView.cpp:1]: There are still other signal/slots connections on this file with the old syntax.

**Review by adriazalvarez — CHANGES_REQUESTED** [review]

**Review by adriazalvarez — APPROVED** [review]
Thanks for the changes, peak fitting is working correctly and the interface organization is much improved.

**Pull request merged**